### PR TITLE
525 more flexibility in the var= operator

### LINKED ISSUE #525: more flexibility in the `var= <<` operator
in https://github.com/ploomber/jupysql/pull/501 `var= <<` was introduced to print the data frame while assigning it. However, it's too strict and we should allow other versions to have the same effect. Essentially, we should ignore the space surrounding the `=`, just like f-strings in Python:【 code pythoh 】

In [4]: x = 'hello'

In [5]: f"{x=}"

Out[5]: "x='hello'"

In [6]: f"{x =}"

Out[6]: "x ='hello'"

In [7]: f"{x = }"

Out[7]: "x = 'hello'"

In [8]: f"{x}"

Out[8]: 'hello'

In [9]: f"{x =  }"

Out[9]: "x =  'hello'"

【 end code 】

### PR DESCRIPTION
**Pull request by tl3119**
## Describe your changes
Add` var = <<` and `var =<<`, increase the flexibility
## Issue number

Closes #525 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--609.org.readthedocs.build/en/609/

### CODE REVIEW

**Comment by tl3119** [comment]
changes on api folder and howto folder is the format update. Changes on parse.py and test_magic.py and test_parse.py are relevant to this issue

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/tests/test_magic.py [hunk 1]
```diff
@@ -111,34 +111,51 @@ def test_result_var_multiline_shovel(ip):
...
-    result = ip.run_cell_magic(
-        "sql",
-        "",
-        """
-        sqlite://
-        x= <<
-        SELECT last_name FROM author;
-        """,
-    )
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (tl3119) [reply src/tests/test_magic.py:114]: I add test case for `  =     << ` and `       =     << `. I also try these command in binder, and all the commands can work

File: doc/api/magic-plot.md [hunk 2]
```diff
@@ -152,10 +152,11 @@ generate histograms without explicitly removing NULL entries.
 `%sqlplot` returns a `matplotlib.Axes` object.
 
 ```{code-cell} ipython3
-ax = %sqlplot histogram --table penguins.csv --column body_mass_g 
+ax = %sqlplot histogram --table penguins.csv --column body_mass_g
```
Comment (neelasha23) [inline doc/api/magic-plot.md:155]: Please revert these changes if not part of the issue

File: doc/api/magic-sql.md [hunk 3]
```diff
@@ -136,8 +136,8 @@ Or pass an alias (**added in 0.5.2**):
 %sql --close db-two
 ```
 
-
 ## Specify creator function
+
```
Comment (neelasha23) [inline doc/api/magic-sql.md:140]: Please revert these changes if not part of the issue

File: doc/howto/format-sql.ipynb [hunk 4]
```diff
@@ -95,7 +95,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.11.0b4"
+   "version": "3.11.0"
```
Comment (neelasha23) [inline doc/howto/format-sql.ipynb:98]: Why is this file added to the PR?

File: doc/howto/ggplot-interact.md [hunk 5]
```diff
@@ -193,8 +193,11 @@ show_legend = widgets.ToggleButton(
 
 ```{code-cell} ipython3
 def plot(b, cmap, show_legend):
-    (ggplot("diamonds", aes(x="price")) + geom_histogram(bins=b, fill="cut", cmap=cmap)
-     + facet_wrap("color", legend=show_legend))
+    (
+        ggplot("diamonds", aes(x="price"))
+        + geom_histogram(bins=b, fill="cut", cmap=cmap)
```
Comment (neelasha23) [inline doc/howto/ggplot-interact.md:198]: Please revert if not part of the issue

**Comment by neelasha23** [comment]
These cases are throwing error. Are we not supporting these?

```
%%sql lang=<<SELECT *
FROM languages
```

```
%%sql lang= <<SELECT *
FROM languages
```

```
%%sql lang=<< SELECT *
FROM languages
```

```
%%sql lang =<<SELECT *
FROM languages
```

```
%%sql lang = <<SELECT *
FROM languages
```

Here's a screenshot:

Screenshot 2023-06-13 at 12 34 07 PM

File: src/sql/parse.py [hunk 6]
```diff
@@ -65,6 +65,16 @@ def parse(cell, config):
         if len(pieces) == 2:
             return result
         cell = pieces[2]
+    elif len(pieces) > 1 and pieces[1] == "=<<":
```
Comment (neelasha23) [inline src/sql/parse.py:68]: Please add some comments here

File: src/tests/test_magic.py [hunk 7]
```diff
@@ -111,34 +111,67 @@ def test_result_var_multiline_shovel(ip):
...
+        ),
+    ],
+)
+def test_return_result_var(ip, sql_statement, expected_result):
+    result = ip.run_cell_magic("sql", "", sql_statement)
     var = ip.user_global_ns["x"]
     assert "Shakespeare" in str(var) and "Brecht" in str(var)
-    assert result.dict() == {"last_name": ("Shakespeare", "Brecht")}
+    if result is None:
+        assert result is None
```
Comment (neelasha23) [inline src/tests/test_magic.py:172]: This can be moved to parametrize

File: src/tests/test_parse.py [hunk 8]
```diff
@@ -85,6 +85,18 @@ def test_parse_return_shovel_operator():
         "result_var": "dest",
         "return_result_var": True,
     }
+    assert parse("dest = << SELECT * FROM work", empty_config) == {
+        "connection": "",
+        "sql": "SELECT * FROM work",
+        "result_var": "dest",
+        "return_result_var": True,
+    }
+    assert parse("dest =<< SELECT * FROM work", empty_config) == {
```
Comment (neelasha23) [inline src/tests/test_parse.py:94]: Please add more test cases here, like:
`dest =        << SELECT * FROM work`
`dest      =<< SELECT * FROM work`
`dest =          << SELECT * FROM work`
etc

File: doc/api/magic-plot.md (hunk 2 above)
Comment (tl3119) [reply doc/api/magic-plot.md]: revert it now

File: doc/howto/format-sql.ipynb (hunk 4 above)
Comment (tl3119) [reply doc/howto/format-sql.ipynb]: revert it

File: doc/howto/ggplot-interact.md (hunk 5 above)
Comment (tl3119) [reply doc/howto/ggplot-interact.md]: revert it

File: src/sql/parse.py (hunk 6 above)
Comment (tl3119) [reply src/sql/parse.py:70]: Add the comments now!

File: src/tests/test_magic.py (hunk 7 above)
Comment (tl3119) [reply src/tests/test_magic.py:172]: which one should i move to parametrize? parametrize already has the result now

File: src/tests/test_parse.py (hunk 8 above)
Comment (tl3119) [reply src/tests/test_parse.py:94]: Sure, add more tests now

**Comment by tl3119** [comment]
> These cases are throwing error. Are we not supporting these?
> 
> ```
> %%sql lang=<<SELECT *
> FROM languages
> ```
> 
> ```
> %%sql lang= <<SELECT *
> FROM languages
> ```
> 
> ```
> %%sql lang=<< SELECT *
> FROM languages
> ```
> 
> ```
> %%sql lang =<<SELECT *
> FROM languages
> ```
> 
> ```
> %%sql lang = <<SELECT *
> FROM languages
> ```
> 
> Here's a screenshot:
> 
> Screenshot 2023-06-13 at 12 34 07 PM

From my previous understanding, the syntax should have at least one space after "<<" to distinguish the syntax between the operator and sql query, so I assume it not support these syntax. But if we want to support these, I can update it.

File: src/tests/test_magic.py (hunk 7 above)
Comment (neelasha23) [reply src/tests/test_magic.py:172]: This statement looks redundant. Why do we need this? :
```
if result is None:
        assert result is None
```
For every case the expected result should be in parametrize and you need to assert `expected_result == actual_result`

**Comment by neelasha23** [comment]
Okay, what about this one:
```
%%sql lang=<< SELECT *
FROM languages
```
This should be supported since it has one space after `<<`

@tl3119

File: src/tests/test_magic.py (hunk 7 above)
Comment (neelasha23) [reply src/tests/test_magic.py:172]: Maybe you can add 

```
if result is not None:
  result = result.dict()
assert result == expected_result
```

File: src/sql/parse.py [hunk 9]
```diff
@@ -65,6 +66,18 @@ def parse(cell, config):
         if len(pieces) == 2:
             return result
         cell = pieces[2]
+    # deal with situation var =<< 'sql'
+    elif len(pieces) > 1 and pieces[1] == "=<<":
+        result["result_var"] = pieces[0]
+        result["return_result_var"] = True
+        cell = pieces[2]
+
+    # deal with situation var = << 'sql'
```
Comment (neelasha23) [inline src/sql/parse.py:75]: `deal with situation var = << 'sql'` -> `handle flexible spacing around =`

File: src/tests/test_parse.py [hunk 10]
```diff
@@ -85,6 +85,36 @@ def test_parse_return_shovel_operator():
...
+        "connection": "",
+        "sql": "SELECT * FROM work",
+        "result_var": "dest",
+        "return_result_var": True,
+    }
+    assert parse("dest =        << SELECT * FROM work", empty_config) == {
+        "connection": "",
+        "sql": "SELECT * FROM work",
+        "result_var": "dest",
```
Comment (neelasha23) [inline src/tests/test_parse.py:103]: This dict seems same in all cases. It can be defined in one place

File: src/tests/test_parse.py [hunk 11]
```diff
@@ -79,42 +79,19 @@ def test_parse_shovel_operator():
...
-        "return_result_var": True,
-    }
-    assert parse("dest      =<< SELECT * FROM work", empty_config) == {
-        "connection": "",
-        "sql": "SELECT * FROM work",
-        "result_var": "dest",
-        "return_result_var": True,
-    }
-    assert parse("dest =          << SELECT * FROM work", empty_config) == {
+    test_dict = {
```
Comment (neelasha23) [inline src/tests/test_parse.py:82]: rename to `result`

File: src/tests/test_magic.py (hunk 7 above)
Comment (tl3119) [reply src/tests/test_magic.py]: Sure, update it now

File: src/sql/parse.py (hunk 9 above)
Comment (tl3119) [reply src/sql/parse.py]: Update it now

File: src/tests/test_parse.py (hunk 11 above)
Comment (tl3119) [reply src/tests/test_parse.py]: update it now

**Comment by tl3119** [comment]
> Okay, what about this one:
> 
> ```
> %%sql lang=<< SELECT *
> FROM languages
> ```
> 
> This should be supported since it has one space after `<<`
> 
> @tl3119

Update it, and this can work now

**Comment by neelasha23** [comment]
Looks good overall! A few more things:
1. Please click the `resolve conversation` button for the ones which have been resolved.
2. Add a changelog entry for this change
3. Add a note in the docs [assignment section](https://jupysql.ploomber.io/en/latest/intro.html#assignment) that we support flexible spacing and add one example.

@tl3119

File: doc/intro.md [hunk 12]
```diff
@@ -161,6 +161,13 @@ returning the results.
 FROM languages
 ```
 
+The `myvar= <<` supports flexible spaceing, such as `myvar = <<` and `myvar =<<`. All these syntaxs can work, for example:
```
Comment (neelasha23) [inline doc/intro.md:164]: `spaceing` -> `spacing`
`syntaxs` -> `syntaxes`

Please reframe to :

`Flexible spacing is supported in assignments, e.g., `myvar = <<` and `myvar =<<`. Let's see an example below:`

```{code-cell} ipython3
%%sql lang = << SELECT *
FROM languages
```

Comment (neelasha23) [reply doc/intro.md:164]: Use can use Grammarly for checking typos and grammatical errors.

Comment (tl3119) [reply doc/intro.md]: Got it, I will be more careful checking those typos and grammatical errors next time.

**Review by neelasha23 — DISMISSED** [review]

File: src/tests/test_parse.py [hunk 13]
```diff
@@ -79,12 +79,19 @@ def test_parse_shovel_operator():
...
         "result_var": "dest",
         "return_result_var": True,
     }
+    assert parse("dest= << SELECT * FROM work", empty_config) == result
+    assert parse("dest = << SELECT * FROM work", empty_config) == result
+    assert parse("dest =<< SELECT * FROM work", empty_config) == result
+    assert parse("dest =        << SELECT * FROM work", empty_config) == result
+    assert parse("dest      =<< SELECT * FROM work", empty_config) == result
+    assert parse("dest =          << SELECT * FROM work", empty_config) == result
+    assert parse("dest=<< SELECT * FROM work", empty_config) == result
```
Comment (edublancas) [inline src/tests/test_parse.py:94]: use pytest.mark.parametrize here instead of having multiple assert statements

**Review by edublancas — CHANGES_REQUESTED** [review]
minor comment, please fix merge conflicts in changelog.md

**Comment by tl3119** [comment]
> minor comment, please fix merge conflicts in changelog.md

Fix now

File: CHANGELOG.md [hunk 14]
```diff
@@ -17,6 +17,7 @@
 * [Doc] Snowflake integration guide ([#384](https://github.com/ploomber/jupysql/issues/384))
 * [Doc] User guide on using JupySQL in `.py` scripts ([#449](https://github.com/ploomber/jupysql/issues/449))
 * [Doc] Added `%magic?` to APIs and quickstart ([#97](https://github.com/ploomber/jupysql/issues/97))
+* [Feature] Support flexible spacing `myvar=<<` operator ([#525](https://github.com/ploomber/jupysql/issues/525))
```
Comment (edublancas) [inline CHANGELOG.md:20]: please move this to the top section. we just made a release and this is no longer the development section

File: doc/intro.md [hunk 15]
```diff
@@ -161,6 +161,13 @@ returning the results.
 FROM languages
 ```
 
+Flexible spacing is supported in assignments, e.g., myvar = << and myvar =<<. Let's see an example below:
+
+```{code-cell} ipython3
+%%sql lang = << SELECT *
+FROM languages
+```
+
```
Comment (edublancas) [inline doc/intro.md:170]: no need to add this to the docs

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/tests/test_parse.py [hunk 16]
```diff
@@ -78,13 +78,26 @@ def test_parse_shovel_operator():
...
+        "dest= << SELECT * FROM work",
+        "dest = << SELECT * FROM work",
+        "dest =<< SELECT * FROM work",
+        "dest =        << SELECT * FROM work",
+        "dest      =<< SELECT * FROM work",
+        "dest =          << SELECT * FROM work",
+        "dest=<< SELECT * FROM work",
+    ],
+)
+def test_parse_return_shovel_operator(input_string):
```
Comment (yafimvo) [inline src/tests/test_parse.py:93]: @edublancas 
Do we allow these cases?

`dest= <<SELECT * FROM work`

`dest<< SELECT * FROM work`

File: src/sql/parse.py [hunk 17]
```diff
@@ -65,6 +71,18 @@ def parse(cell, config):
...
+        result["result_var"] = pieces[0]
+        result["return_result_var"] = True
+        cell = pieces[2]
+
+    # handle flexible spacing around =
+    pieces = cell.split(None, 3)
+    if len(pieces) > 1 and pieces[1] == "=" and pieces[2] == "<<":
+        result["result_var"] = pieces[0]
+        result["return_result_var"] = True
+        cell = pieces[3]
```
Comment (yafimvo) [inline src/sql/parse.py:85]: Is it possible to generalize these 3 code blocks? the logic seems to repeat itself.

File: src/tests/test_parse.py (hunk 16 above)
Comment (yafimvo) [reply src/tests/test_parse.py:93]: I think we can add another test for more complex queries such:

```
%sql --save save_result result<< SELECT * FROM work limit 2
```

This query should store `save_result` and `result`.

(it works)

**Review by yafimvo — CHANGES_REQUESTED** [review]
added some notes

File: CHANGELOG.md (hunk 14 above)
Comment (tl3119) [reply CHANGELOG.md]: move it to the top section now

File: doc/intro.md (hunk 15 above)
Comment (tl3119) [reply doc/intro.md]: delete it now

File: src/tests/test_parse.py (hunk 16 above)
Comment (tl3119) [reply src/tests/test_parse.py]: I assume we do not allow the first one, since we need at least one space after `<<`

Comment (tl3119) [reply src/tests/test_parse.py]: > I think we can add another test for more complex queries such:
> 
> ```
> %sql --save save_result result<< SELECT * FROM work limit 2
> ```
> 
> This query should store `save_result` and `result`.
> 
> (it works)

What's the meaning of this test, since it seems does not have the `=<<` operator used in this PR, only `<<` operator

**Comment by tl3119** [comment]
The test pass locally in my computer, but fails in github, what's the reason?
image

**Review by edublancas — APPROVED** [review]

**Comment by edublancas** [comment]
great!
